### ISSUE TITLE
[Debugger] Ability to add debugger breakpoints inside editor text/source files

### ISSUE THREAD

**Issue by bobcolner** [jupyterlab/jupyterlab#9525]
### Problem

Not able to use debugger functionality inside jupyterlab editor files.

### Proposed Solution

Add ability to add breakpoint to editor files. The source could be run from the console  or notebook (that imported the relevant source file)

**Comment by jtpio** [comment]
Thanks @bobcolner for opening this.

It should already be possible to debug files, but the workflow might not be the most intuitive yet as of today and could indeed be improved.

For now the file needs to be backed by a kernel using the "Create Console for Editor": 

![debug-files-lab-3](https://user-images.githubusercontent.com/591645/103551988-7c7ef000-4eab-11eb-8d8f-f0aaef0e1091.gif)

**Comment by bobcolner** [comment]
Thanks for the interest.

my common (amateur?) workflow is basically using the notebook as my dev interface to a library of custom modules that are always evolving. 

Ideally I could add a breakpoint in the editor UI and be able to access the debugger after triggering the breakpoint from a notebook cell (hope that makes sense).

**Comment by CarlAndersson** [comment]
Following the gif above, it's only possible to debug a "script". If I have code in a local package/module that needs debugging, I would need to place a breakpoint inside a function which is in a source file. Since the function call will not be in the same file as the definition of the function, I cannot get this to work.

Tried the following:
- Creating a console for the source file, importing the package, and running the function from the console.
- Creating another script which imports the package, and runs the function. Launching a console for the source file and connecting it to the same kernel as the script, then running the script.
Both of these simply ignore the breakpoint in the function I would like to debug...

**Comment by krassowski** [comment]
My (different than OP) use case would be to debug a module that I have outside of notebook, but which is called and imported from within the notebook. The existing notebook kernel is already executing this code so no need to create a new kernel; the issue is that I cannot set a breakpoint for this module. I need to set it in the notebook and then step into the function of interest which is sub-optimal.

Is this already possible somehow, am I missing something maybe?

**Comment by jtpio** [comment]
> Is this already possible somehow, am I missing something maybe?

For now there needs to be a live kernel associated to the file being opened. The main reason is that setting a breakpoint requires sending a debug message to the kernel. There is also the case of remote kernels where the files and notebooks might not be visible by the kernel.

There is an item in https://github.com/jupyterlab/jupyterlab/issues/10139#issue-865417157 that is about setting breakpoints in any file (loaded source), similar to the browser dev tools.

**Comment by krassowski** [comment]
What is stopping it from working when I point the session to the same kernel as does my notebook?

![test-fail](https://user-images.githubusercontent.com/5832902/122084825-12735b00-cdfa-11eb-8f9f-4835beb05d48.gif)

**Comment by pmarcellino** [comment]
Yes @krassowski this would be the perfect functionality. I did check it's not an import order issue. When replicating what you did but importing after adding the breakpoint that doesn't help.

Somehow if you put a break point at the line calling the function, and then step in you can add a break point in the source file there, but I don't know why that works.
![image](https://user-images.githubusercontent.com/46568443/122100340-af250100-cdd8-11eb-8997-9a52f197c078.png)

**Comment by jtpio** [comment]
> What is stopping it from working when I point the session to the same kernel as does my notebook?

I would have expected this to work yes.

**Comment by pmarcellino** [comment]
Thinking about it, although it is less intuitive, having the functionality be that you have to point the kernal at the file to insert the break point would be better functionality, since the same file could actively be being use by multiple kernals at the same time.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jupyterlab-ide-customization/11271/4
